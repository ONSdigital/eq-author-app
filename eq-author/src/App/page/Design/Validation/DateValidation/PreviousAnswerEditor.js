import React from "react";
import PropTypes from "prop-types";
import { get } from "lodash";

import PreviousAnswerContentPicker from "../PreviousAnswerContentPicker";

const PreviousAnswerEditor = ({ onChangeUpdate, answer, validation, readKey, }) => (
  <PreviousAnswerContentPicker
    answerId={answer.id}
    onSubmit={onChangeUpdate}
    selectedContentDisplayName={get(validation.previousAnswer, "displayName")}
    selectedId={get(validation.previousAnswer, "id")}
    path={`answer.validation.${readKey}.availablePreviousAnswers`}
    data-test="content-picker-select"
  />
);

PreviousAnswerEditor.propTypes = {
  validation: PropTypes.shape({
    previousAnswer: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
  }).isRequired,
  answer: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }).isRequired,
  readKey: PropTypes.string.isRequired,
  onChangeUpdate: PropTypes.func.isRequired,
};

export default PreviousAnswerEditor;
