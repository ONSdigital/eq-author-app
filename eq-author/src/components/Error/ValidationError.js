import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { colors } from "constants/theme";
import VisuallyHidden from "../VisuallyHidden/index";

import IconText from "components/IconText";
import WarningIcon from "constants/icon-warning.svg?inline";

const ErrorStyling = styled(IconText)`
  color: ${colors.red};
  width: ${(props) => (props.right ? "100%" : "80%")};
  justify-content: flex-end;
  margin: 0.5em 0;
`;

export const ValidationError = (props) => {
  const { children, right, className } = props;

  return (
    <ErrorStyling className={className} right={right} icon={WarningIcon}>
      <VisuallyHidden>Error:&nbsp;</VisuallyHidden>
      {children}
    </ErrorStyling>
  );
};

ValidationError.propTypes = {
  children: PropTypes.node.isRequired,
  right: PropTypes.bool,
  className: PropTypes.string,
};
