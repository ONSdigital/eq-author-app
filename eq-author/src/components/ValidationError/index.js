import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { colors } from "constants/theme";

import IconText from "components/IconText";
import WarningIcon from "constants/icon-warning.svg?inline";
import VisuallyHidden from "../VisuallyHidden/index";

const ErrorStyling = styled(IconText)`
  color: ${colors.red};
  width: ${props => (props.right ? "100%" : "80%")};
  justify-content: flex-end;
  margin: 0.5em 0;
`;

const ValidationError = ({ children, right, className, test }) => (
  <ErrorStyling
    icon={WarningIcon}
    className={className}
    right={right}
    data-test={test}
  >
    <VisuallyHidden>Error:&nbsp;</VisuallyHidden>
    {children}
  </ErrorStyling>
);

ValidationError.propTypes = {
  children: PropTypes.node.isRequired,
  right: PropTypes.bool,
  className: PropTypes.string,
  test: PropTypes.string,
};

export default ValidationError;
