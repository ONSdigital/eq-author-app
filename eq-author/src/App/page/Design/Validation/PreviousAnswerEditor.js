import React from "react";
import PropTypes from "prop-types";
import ValidationError from "components/ValidationError";

import FieldWithInclude from "./FieldWithInclude";
import * as entityTypes from "constants/validation-entity-types";
import PreviousAnswerContentPicker from "./PreviousAnswerContentPicker";

import {
  ERR_REFERENCE_DELETED,
  ERR_REFERENCE_MOVED,
} from "constants/validationMessages";

export const errorMessages = {
  ERR_NO_VALUE: "Answer is required",
  ERR_REFERENCE_DELETED,
  ERR_REFERENCE_MOVED,
};

const PreviousAnswerEditor = ({
  onChangeUpdate,
  validation,
  answer: { type },
}) => {
  const { errors } = validation.validationErrorInfo;

  const errorCode = errors
    .map((e) => e.errorCode)
    .find((errorCode) => Object.keys(errorMessages).includes(errorCode));

  const contentPicker = (
    <PreviousAnswerContentPicker
      onSubmit={onChangeUpdate}
      selectedContentDisplayName={validation?.previousAnswer?.displayName}
      selectedId={validation?.previousAnswer?.id}
      data-test="content-picker-select"
      hasError={Boolean(errorCode)}
      allowedAnswerTypes={[type]}
    />
  );

  const component = validation?.__typename?.includes("Date") ? (
    contentPicker
  ) : (
    <FieldWithInclude
      id="inclusive"
      name="inclusive"
      onChange={onChangeUpdate}
      checked={validation.inclusive}
    >
      {contentPicker}
    </FieldWithInclude>
  );

  return (
    <>
      {component}
      {errorCode && (
        <ValidationError>{errorMessages[errorCode]}</ValidationError>
      )}
    </>
  );
};

PreviousAnswerEditor.propTypes = {
  validation: PropTypes.shape({
    id: PropTypes.string.isRequired,
    enabled: PropTypes.bool.isRequired,
    custom: PropTypes.number,
    inclusive: PropTypes.bool.isRequired,
    previousAnswer: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
    entityType: PropTypes.oneOf(Object.values(entityTypes)),
  }).isRequired,
  readKey: PropTypes.string.isRequired,
  onChangeUpdate: PropTypes.func.isRequired,
  answer: PropTypes.shape({
    type: PropTypes.string.isReqeuired,
  }).isRequired,
};

export default PreviousAnswerEditor;
