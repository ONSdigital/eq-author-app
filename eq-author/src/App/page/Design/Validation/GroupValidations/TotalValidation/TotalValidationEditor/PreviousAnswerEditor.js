import React from "react";
import PropTypes from "prop-types";
import { get } from "lodash";

import { Field } from "components/Forms";
import ValidationError from "components/ValidationError";

import { usePage } from "components/QuestionnaireContext";

import {
  ERR_REFERENCE_DELETED,
  ERR_REFERENCE_MOVED,
} from "constants/validationMessages";

import PreviousAnswerContentPicker from "../../../PreviousAnswerContentPicker";

export const errorMessages = {
  ERR_NO_VALUE: "Answer required",
  ERR_REFERENCE_DELETED,
  ERR_REFERENCE_MOVED,
};

const PreviousAnswerEditor = ({ total, onChangeUpdate, errors }) => {
  const error = errors?.[0];

  const page = usePage();
  const allowedAnswerType = page?.answers?.[0]?.type;

  return (
    <Field>
      <PreviousAnswerContentPicker
        onSubmit={onChangeUpdate}
        selectedContentDisplayName={get(total.previousAnswer, "displayName")}
        data-test="content-picker-select"
        allowedAnswerTypes={[allowedAnswerType]}
        hasError={Boolean(error)}
      />
      {error && (
        <ValidationError right={false}>
          {errorMessages[error.errorCode]}
        </ValidationError>
      )}
    </Field>
  );
};

PreviousAnswerEditor.propTypes = {
  total: PropTypes.shape({
    previousAnswer: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
  }).isRequired,
  onChangeUpdate: PropTypes.func.isRequired,
  errors: PropTypes.arrayOf(
    PropTypes.shape({
      errorCode: PropTypes.string,
      field: PropTypes.string,
      id: PropTypes.string,
      type: PropTypes.string,
    })
  ),
};

export default PreviousAnswerEditor;
