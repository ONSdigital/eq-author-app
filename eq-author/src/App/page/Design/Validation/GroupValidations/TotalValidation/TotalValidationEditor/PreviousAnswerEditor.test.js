import React from "react";
import { shallow } from "enzyme";

import PreviousAnswerEditor from "./PreviousAnswerEditor";
import ValidationError from "components/ValidationError";

describe("TotalValidation: Previous Answer Editor", () => {
  let props;
  beforeEach(() => {
    props = {
      total: {
        id: "1",
        previousAnswer: {
          displayName: "My answer",
        },
      },
      onChangeUpdate: jest.fn(),
    };
  });

  const render = () => shallow(<PreviousAnswerEditor {...props} />);

  it("should render", () => {
    expect(render()).toMatchSnapshot();
  });

  it("should trigger onChange when the number input changes", () => {
    const wrapper = render();
    wrapper
      .find("[data-test='content-picker-select']")
      .simulate("submit", { name: "previousAnswer", value: { id: "1" } });
    expect(props.onChangeUpdate).toHaveBeenCalledWith({
      name: "previousAnswer",
      value: { id: "1" },
    });
  });

  it("should display validation errors if set", () => {
    props.errors = [
      {
        id: "err-1",
        errorCode: "ERR_NO_VALUE",
      },
    ];

    const wrapper = render();
    expect(wrapper.find(ValidationError)).toHaveLength(1);
  });
});
