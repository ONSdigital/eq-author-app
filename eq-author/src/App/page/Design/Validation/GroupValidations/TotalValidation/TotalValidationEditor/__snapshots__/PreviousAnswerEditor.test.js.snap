// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TotalValidation: Previous Answer Editor should render 1`] = `
<Field
  disabled={false}
  last={false}
>
  <PreviousAnswerContentPicker
    allowedAnswerTypes={
      Array [
        undefined,
      ]
    }
    data-test="content-picker-select"
    hasError={false}
    onSubmit={[MockFunction]}
    selectedContentDisplayName="My answer"
  />
</Field>
`;
